packages/modal-core/src/utils/CoreUtil.ts

```diff
@@ -59,6 +71,23 @@ export const CoreUtil = {
       return this.formatNativeUrl(appUrl, wcUri, name)
     }
     let safeAppUrl = appUrl
+    //  Universal link required in telegram context
+    if (safeAppUrl.startsWith('https://t.me')) {
+      // eslint-disable-next-line require-unicode-regexp
+      const formattedUri = Buffer.from(wcUri).toString('base64').replace(/[=]/g, '')
+      if (safeAppUrl.endsWith('/')) {
+        safeAppUrl = safeAppUrl.slice(0, -1)
+      }
+
+      this.setWalletConnectDeepLink(safeAppUrl, name)
+
+      const link = safeAppUrl.includes('?')
+        ? `${safeAppUrl}&startapp=${formattedUri}`
+        : `${safeAppUrl}?startapp=${formattedUri}`
```

Comment:
For more safer and readable approach, we could use `URL` class to manipulate search params:

```suggestion
      const url = new URL(safeAppUrl);
      url.searchParams.set('startapp', formattedUri);
      const link = url.toString();
```